Book and chapter selector UX improvements

jmarsican opened the pull request:
**Book and chapter selector user experience improvements for easy and fast navigation among passages**
- Use modern look & feel for buttons with default material design instead of gradient
- Show full book name hint
- Transpose books buttons table grid
- Use slightly different color for new testament buttons to reduce visual effort while navigating among passages
- Improve book preview hint text
- Tweaks and fixes for #332 #142 #715 

**Screenshots**

![finalLands](https://user-images.githubusercontent.com/10614631/116013370-42765d00-a606-11eb-90fa-af14935380ee.png)

![finalBook](https://user-images.githubusercontent.com/10614631/116012884-a21f3900-a603-11eb-971d-e22aed829f14.png)

Commit:
Book and chapter selector UX improvements

- Use modern look & feel for buttons with default material design instead of gradient
- Show full book name hint
- Transpose books buttons table grid

Commit:
- Set new testament buttons with different color to reduce search effort

Commit:
- Improve book preview hint text

tuomas2 commented:
Looks good! Checking this more carefully on next thursday.

timbze commented:
Very nice UI improvements! I have a few suggestions, but @tuomas2 should affirm or deny my suggestions before you make changes based on them.

- [ ] Leave the default book order the way it was (in portrait mode the books go by column, in landscape by row)
- [x] Add back (up) button in toolbar (can be done by modifying manifest file) #997
- [ ] All settings would save in Db that way the user's preference would again be used next time
- [ ] Instead of current toolbar sort button on right, add a 3-dot menu with:
  * Sort by book order [checkbox in menu group 1] << this would be default
  * Sort alphabetically [checkbox in menu group 1]
  * Separator
  * Order books by row [checkbox in menu group 2] << off by default, but when checked the books would also order by row in portrait mode the way landscape already does by default

tuomas2 commented:
Good suggestions @timbze. Settings should be to shared preferences (CommonUtils.sharedPreferences).

tuomas2 commented:
@timbze problem with "back" button is that this activity is used from many places (not just mainbibleactivity, but at least from SpeakActivity as well)

timbze commented:
> @timbze problem with "back" button is that this activity is used from many places (not just mainbibleactivity, but at least from SpeakActivity as well)

Ok in that case it would need to be manually implemented to run "back" action right? I think I've done it with a few lines of code.

I just think the screen would look more user friendly if it had the back button

tuomas2 commented:
> > @timbze problem with "back" button is that this activity is used from many places (not just mainbibleactivity, but at least from SpeakActivity as well)
> 
> Ok in that case it would need to be manually implemented to run "back" action right? I think I've done it with a few lines of code.

Code example for this would be nice. I'm in the impression that it's not very convenient to override default behavior. (Kind of remember trying this at some point)

> I just think the screen would look more user friendly if it had the back button

I agree.

timbze commented:
EDIT: Made PR #997

> Code example for this would be nice.

See https://github.com/timbze/and-bible/tree/demo/implement_back_action
* Code back (up) button for GridChoosePassageBook: https://github.com/timbze/and-bible/commit/0af9bd9b2a3d89b0f709101881b072db9bf627da
* Manifest.xml back (up) for GridChoosePassageChapter and Verse https://github.com/timbze/and-bible/commit/1f82066b329e0843d494ae0c32b70054e71305c6

@tuomas2 if you see something wrong with this or something that might not work let me know. I have used this type of code in my personal apps.

jmarsican commented:
Glad you liked these improvements folks. The purpose of this PR was to address items specified in #332 as per @tuomas2 requested in my previous PR #988.  I like suggestions @timbze made too, but perhaps they are out of scope and might be addressed in another PR. 
I can revert changes to traspose books grid order and would try to investigate how to implement such 3-dot menu button with those three sort/ordering strategies. Although, it might take long time to understand and design implementation (I don't know how local storage might be implemented) and I will not be able to do it before next weekend.

timbze commented:
@jmarsican thanks for your help in contributing to the app, we're always glad for contributors!

Pull requests are usually not a rush, we work on stuff as we have time.

I can give you pointers on how to implement the 3-dot menu and saving of settings, it's really not that hard since most hard work in the code is already there.

I wouldn't mind if the other things were in a separate PR, but the default order needs to stay the way it was, so if you revert that then feature-wise IMO this PR would be ready to merge.

I have at least one code suggestion. Will look over it again and submit code reviews if I have any other.

timbze commented on app/src/main/java/net/bible/android/view/activity/navigation/GridChoosePassageBook.kt:
```diff
@@ -47,12 +47,14 @@ import java.util.ArrayList
 
 import javax.inject.Inject
 
+private val NEW_TESTAMENT_TINT = Color.parseColor("#FF505050")
```
Variables should generally be in a class. My suggestion is to move this to the area where `isCurrentlyShowingScripture` is.

timbze commented on app/src/main/java/net/bible/android/view/activity/navigation/GridChoosePassageChapter.kt:
```diff
@@ -103,6 +103,7 @@ class GridChoosePassageChapter : CustomTitlebarActivityBase(), OnButtonGridActio
             // this is used for preview
             buttonInfo.id = i
             buttonInfo.name = Integer.toString(i)
+            buttonInfo.description = Integer.toString(i)
```
Should be `i.toString()`. Also the line above it.

jmarsican commented:
> @jmarsican thanks for your help in contributing to the app, we're always glad for contributors!
> 
> Pull requests are usually not a rush, we work on stuff as we have time.
> 
> I can give you pointers on how to implement the 3-dot menu and saving of settings, it's really not that hard since most hard work in the code is already there.
> 
> I wouldn't mind if the other things were in a separate PR, but the default order needs to stay the way it was, so if you revert that then feature-wise IMO this PR would be ready to merge.
> 
> I have at least one code suggestion. Will look over it again and submit code reviews if I have any other.

For sure Timmy. I am going to restore default order the way it was shortly. 

Please do share any clue on how to implement saving settings so I can analyze and implement as soon as I can.

Looking forward to get any code suggestions that might improve the code of this PR.

Best Regards.

tuomas2 commented:
Most recent example for saving sharedPreferences:
ShareWidget.kt 
reading on line 44
saving on line 65

timbze commented on app/src/main/java/net/bible/android/view/util/buttongrid/ButtonGrid.kt:
```diff
@@ -218,14 +212,13 @@ class ButtonGrid constructor(context: Context, attrs: AttributeSet? = null, defS
+                Log.d(TAG, "Previewing " + buttonInfo.description)
                 currentPreview = buttonInfo
-                previewText!!.text = buttonInfo.name
+                previewText!!.text = buttonInfo.description
```
The double exclamation `!!` should not be used normally. It causes a crash when there's null, and especially in cases like this where it's not some important app logic, it should never cause a crash (even though it should never be null). I see you just used what was there from before and that's not a problem 😄 

It would be better to add a new line under `Log.d` (line 216), like this:
`val preview = previewText ?: return`

And then now use `preview.` everywhere underneath where there previously was `previewText!!.`

jmarsican replied:
Yes my bad. I use to follow boy scout rule but I forgot for this class. Looks like it was a Java class automatically converted to Kotlin and there are plenty of Java-ish remains to be reviewes :sweat_smile:

timbze commented on app/src/main/java/net/bible/android/view/util/buttongrid/ButtonGrid.kt:
```diff
@@ -218,14 +212,13 @@ class ButtonGrid constructor(context: Context, attrs: AttributeSet? = null, defS
                 val popupHeight = previewHeight
                 previewText!!.measure(MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED), MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED))
-                val popupWidth = Math.max(previewText!!.measuredWidth, buttonInfo.button!!.width + previewText!!.paddingLeft + previewText!!.paddingRight)
+                val popupWidth = max(previewText!!.measuredWidth, buttonInfo.button!!.width + previewText!!.paddingLeft + previewText!!.paddingRight)
```
Android Studio shows the `!!` after `.button` is useless, you can remove that as well

Commit:
Code review changes

- Restore previous book selectors order
- Move constant inside class
- Fix wrong conversion from Java

jmarsican commented:
@timbze all code improvements you suggested were resolved.
@tuomas2 I've just restored default book order the way it was. Also I reported in #1013 feature to allow user chose preferred order, I'll work on it next weekend.

tuomas2 reviewed (approved):
Looks good, thank you! :-)
